### PR TITLE
chore(sdk): `ChainSpec` generic over hardforks

### LINKED ISSUE #10468: Make `reth_chainspec::ChainSpec` generic over hard fork
### Describe the feature

Make `ChainSpec` generic over the hard fork type. This eventually allows removing optimism feature from `reth-chainspec`.

Many OP feature gated methods on `ChainSpec`, can be sunk a level to hard forks.

### Additional context

Makes sense to re-model `OptimismHardfork` type to have a variant `OptimismHardfork::Eth(EthereumHardfork)`, to reuse logic implemented on `EthereumHardfork`.

_No response_

### PR DESCRIPTION
**Pull request by emhane**
Closes https://github.com/paradigmxyz/reth/issues/10468

Makes `ChainSpec` generic over hard forks, leaving `ChainHardforks` as default generic.

### CODE REVIEW

File: crates/chainspec/src/api.rs [hunk 1]
```diff
@@ -1,7 +1,9 @@
...
-use alloy_chains::Chain;
 use core::fmt::Debug;
 
+use alloy_chains::Chain;
+
+use crate::ChainSpec;
+
 /// Trait representing type configuring a chain spec.
 pub trait EthChainSpec: Send + Sync + Unpin + Debug + 'static {
     // todo: make chain spec type generic over hardfork
```
[the remark this answers is not in the document]
Comment (emhane) [reply crates/chainspec/src/api.rs:9]: thanks :)

**Comment by klkvr** [comment]
hmm, wondering what's the plan here? reuse chainspec with different generic for optimism? what would that generic look like?

**Comment by emhane** [comment]
> hmm, wondering what's the plan here? reuse `ChainSpec` with different generic for optimism? what would that generic look like?

yepp. and any rollup with same chain spec type asides hard forks. it's a building block, to make it easier for rollups to enjoy the customisability of the new reth sdk.

**Comment by mattsse** [comment]
hmm, the chainspec already has some additional chain specific internals, like deposit contract and for other chains it's likely that some additional fields should be included there as well.

so it's imo easier to roll your own, like we already prepped with `OpChainspec`

**Comment by klkvr** [comment]
> > hmm, wondering what's the plan here? reuse `ChainSpec` with different generic for optimism? what would that generic look like?
> 
> yepp. and any rollup with same chain spec type asides hard forks. it's a building block, to make it easier for rollups to enjoy the customisability of the new reth sdk.

makes sense! 

I've changed `EthereumHardforks` trait a bit in https://github.com/paradigmxyz/reth/pull/10978, because it seemed reasonable to move `final_paris_total_difficulty` fn to it. because of that, we now have to implement/require trait on `ChainSpec` directly which I think breaks this usecase a bit. should we consider moving `paris_block_and_final_difficulty` to `ChainHardforks` to support this?

**Comment by emhane** [comment]
> > > hmm, wondering what's the plan here? reuse `ChainSpec` with different generic for optimism? what would that generic look like?
> > 
> > 
> > yepp. and any rollup with same chain spec type asides hard forks. it's a building block, to make it easier for rollups to enjoy the customisability of the new reth sdk.
> 
> makes sense!
> 
> I've changed `EthereumHardforks` trait a bit in #10978, because it seemed reasonable to move `final_paris_total_difficulty` fn to it. because of that, we now have to implement/require trait on `ChainSpec` directly which I think breaks this usecase a bit. should we consider moving `paris_block_and_final_difficulty` to `ChainHardforks` to support this?

let's move it to `Hardforks` trait. it's supertrait of `EthereumHardforks`.

**Comment by emhane** [comment]
> hmm, the chainspec already has some additional chain specific internals, like deposit contract and for other chains it's likely that some additional fields should be included there as well.
> 
> so it's imo easier to roll your own, like we already prepped with `OpChainspec`

it's written as a bullet point in your notion sdk notes?

**Review by mattsse — COMMENTED** [review]
we're going with OpChainSpec and EthChainSpec as standalone types, this is more flexible than trying to unify these in a single generic ChainSpec type moving forward

**Pull request closed without merging**